### ISSUE TITLE
add TypeScript declaration file

### ISSUE THREAD

**Issue by brettz9** [eslint-community/eslint-plugin-promise#488]
### Description

Add a TypeScript declaration file.

### Steps to Reproduce

1.  Add `import promise from 'eslint-plugin-promise'`
2. Hover over `eslint-plugin-promise` in VSC

**Expected behavior:** [What you expect to happen]

Get no errors.

**Actual behavior:** [What actually happens]

Get errors in VSC about TS types

**Comment by brettz9** [comment]
Should we go all in with `checkJs: true`, basically going full JavaScript+JSDoc-as-TypeScript?

**Comment by voxpelli** [comment]
Real life example from neostandard: https://github.com/neostandard/neostandard/blob/08ae5469c71def357f0202fed2be9b83f62efc5a/lib/configs/base.js#L5

> Should we go all in with `checkJs: true`, basically going full JavaScript+JSDoc-as-TypeScript?

I have a PR for that in another project, so I’m +1 on it

**Comment by brettz9** [comment]
Will probably wait on #484 to begin any work here, as don't want the tooling to have to be outdated to be supported...

**Comment by voxpelli** [comment]
Agree @brettz9 👍 And I think @scagood is also interested in this topic :)

**Comment by rgant** [comment]
What is the status of this work? Is it still being considered or will it not be added at this time?

**Comment by JounQin** [comment]
Or even better: rewrite with TypeScript.

**Comment by Aesthermortis** [comment]
Flat config esm

index.d.ts
```javascript
import type { Linter, Rule } from "eslint";

declare const promisePlugin: {
  meta: {
    name: "eslint-plugin-promise";
    namespace: "promise";
    version: string;
  };
  configs: {
    recommended: Linter.Config & {
      name: "promise/recommended";
    };
  };
  rules: Record<string, Rule.RuleModule>;
};

export type PromisePlugin = typeof promisePlugin;
export default promisePlugin;
```

index.js
```javascript
import packageJson from "../package.json" with { type: "json" };
import recommendedRules from "./configs/recommended.js";
import rules from "./rules/index.js";

const { name: packageName, version: packageVersion } = packageJson;
const namespace = "promise";

/** @type {Record<string, import("eslint").Linter.Config>} */
const configs = {};

const plugin = {
  meta: {
    name: packageName,
    namespace,
    version: packageVersion,
  },
  configs,
  rules,
};

Object.assign(plugin.configs, {
  recommended: {
    name: `${namespace}/recommended`,
    plugins: { [namespace]: plugin },
    rules: recommendedRules,
  },
});

export default plugin;
```

**Comment by rgant** [comment]
@voxpelli it would be really great if eslint-plugin-promise could support TypeScript types.

**Comment by Nick2bad4u** [comment]
> @voxpelli it would be really great if eslint-plugin-promise could support TypeScript types.

Can't someone make a @types/ package? 

Should be really easy lol. Idk why this is has been open for 2 years 

If the author wants a PR I can do it .

**Comment by JoshuaKGoldberg** [comment]
> make a @types/ package?

In the meantime, yes, I did - https://github.com/DefinitelyTyped/DefinitelyTyped/pull/75007. Anybody want to take a look? 🙂

**Comment by Nick2bad4u** [comment]
> > make a @types/ package?
> 
> In the meantime, yes, I did - https://github.com/DefinitelyTyped/DefinitelyTyped/pull/75007. Anybody want to take a look? 🙂 

Thank you!!!! 😅😅

**Comment by janpe** [comment]
> > make a @types/ package?
> 
> In the meantime, yes, I did - [DefinitelyTyped/DefinitelyTyped#75007](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/75007). Anybody want to take a look? 🙂

@JoshuaKGoldberg would you mind opening a similar PR for this repository?

**Comment by JoshuaKGoldberg** [comment]
Sure, can do. I just didn't want to trample over @voxpelli's #512. 

How auto-generated would you like the types here to be? The same manually created `.d.ts` would be most straightforward, but then we'd run the risk of getting out of sync with new rules. (which, admittedly, is already an issue with the DefinitelyTyped `@types/` package...)

**Comment by janpe** [comment]
> Sure, can do. I just didn't want to trample over [@voxpelli](https://github.com/voxpelli)'s [#512](https://github.com/eslint-community/eslint-plugin-promise/pull/512).
> 
> How auto-generated would you like the types here to be? The same manually created `.d.ts` would be most straightforward, but then we'd run the risk of getting out of sync with new rules. (which, admittedly, is already an issue with the DefinitelyTyped `@types/` package...)

That PR seems to have been an abandoned draft for almost two years so I think it’s fine to work on a new one
